File: netlify/functions/stripe-webhook.py
"""
Stripe Webhook Handler for CELPIP Webinar Email Delivery
Netlify Function version - deployed automatically with your site

Environment variables needed in Netlify dashboard:
- GMAIL_USER
- GMAIL_PASSWORD (16-char Gmail App Password)
- STRIPE_WEBHOOK_SECRET
- STRIPE_API_KEY
"""

import os
import json
import smtplib
import stripe
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart

# Configuration
STRIPE_API_KEY = os.environ.get("STRIPE_API_KEY")
STRIPE_WEBHOOK_SECRET = os.environ.get("STRIPE_WEBHOOK_SECRET")
GMAIL_USER = os.environ.get("GMAIL_USER")
GMAIL_PASSWORD = os.environ.get("GMAIL_PASSWORD")
GOOGLE_MEET_LINK = "https://meet.google.com/hcf-iwcn-syx"

stripe.api_key = STRIPE_API_KEY


def send_email(customer_email: str, customer_name: str) -> bool:
    """Send confirmation email with Google Meet link."""
    try:
        subject = "Your CELPIP Weekly Webinar Link – Saturday 6:00 PM PST"
        
        body = f"""Hi {customer_name},

Thank you for registering for CELPIP Weekly Webinars!

🗓️ **Session Details:**
Saturday at 6:00 PM PST

📹 **Join on Google Meet:**
{GOOGLE_MEET_LINK}

No account needed—just click the link and join!

**What to expect:**
- 45 minutes of focused CELPIP teaching
- 15 minutes of live Q&A
- Practical strategies you can use immediately

See you Saturday!

—IELTS Corner
hello@ieltscorner.ca"""

        msg = MIMEMultipart()
        msg["From"] = GMAIL_USER
        msg["To"] = customer_email
        msg["Subject"] = subject
        msg.attach(MIMEText(body, "plain"))

        server = smtplib.SMTP_SSL("smtp.gmail.com", 465)
        server.login(GMAIL_USER, GMAIL_PASSWORD)
        server.send_message(msg)
        server.quit()

        print(f"✓ Email sent to {customer_email}")
        return True

    except Exception as e:
        print(f"✗ Failed to send email: {str(e)}")
        return False


def get_customer_details(checkout_session_id: str):
    """Retrieve customer email and name from Stripe checkout session."""
    try:
        session = stripe.checkout.Session.retrieve(checkout_session_id)
        
        customer_email = session.customer_details.email if session.customer_details else None
        customer_name = session.customer_details.name if session.customer_details else "there"
        
        if not customer_email:
            return None
            
        return {
            "email": customer_email,
            "name": customer_name
        }
    except Exception as e:
        print(f"✗ Failed to retrieve session: {str(e)}")
        return None


def handler(event, context):
    """Netlify Function handler for Stripe webhooks."""
    
    # Only accept POST requests
    if event["httpMethod"] != "POST":
        return {
            "statusCode": 405,
            "body": json.dumps({"error": "Method not allowed"})
        }
    
    # Get webhook payload and signature
    payload = event["body"]
    sig_header = event["headers"].get("stripe-signature")
    
    if not sig_header:
        return {
            "statusCode": 400,
            "body": json.dumps({"error": "Missing Stripe signature"})
        }
    
    try:
        # Verify webhook signature
        stripe_event = stripe.Webhook.construct_event(
            payload, sig_header, STRIPE_WEBHOOK_SECRET
        )
    except ValueError:
        return {
            "statusCode": 400,
            "body": json.dumps({"error": "Invalid payload"})
        }
    except stripe.error.SignatureVerificationError:
        return {
            "statusCode": 400,
            "body": json.dumps({"error": "Invalid signature"})
        }
    
    # Handle checkout.session.completed event
    if stripe_event["type"] == "checkout.session.completed":
        checkout_session = stripe_event["data"]["object"]
        
        print(f"New registration: {checkout_session['id']}")
        
        # Only process if payment was successful
        if checkout_session["payment_status"] == "paid":
            customer_info = get_customer_details(checkout_session["id"])
            
            if customer_info:
                success = send_email(customer_info["email"], customer_info["name"])
                
                if success:
                    return {
                        "statusCode": 200,
                        "body": json.dumps({"status": "success", "message": "Email sent"})
                    }
                else:
                    return {
                        "statusCode": 500,
                        "body": json.dumps({"error": "Failed to send email"})
                    }
            else:
                return {
                    "statusCode": 400,
                    "body": json.dumps({"error": "Could not retrieve customer info"})
                }
    
    # Acknowledge receipt of other event types
    return {
        "statusCode": 200,
        "body": json.dumps({"status": "received"})
    }

File: netlify/functions/stripe-webhook.js
import Stripe from 'stripe';
import nodemailer from 'nodemailer';

const GOOGLE_MEET_LINK = 'https://meet.google.com/hcf-iwcn-syx';

function buildEmailHtml(customerName, gmailUser) {
  return `<!DOCTYPE html>
<html>
<head>
  <meta charset="UTF-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
</head>
<body style="margin:0;padding:0;font-family:Arial,sans-serif;background:#f6f4f3;">
  <table role="presentation" style="width:100%;border-collapse:collapse;">
    <tr>
      <td align="center" style="padding:16px;">
        <table role="presentation" style="width:100%;max-width:640px;background:#fff;border:1px solid #e9e1dc;border-radius:10px;overflow:hidden;">
          <tr>
            <td style="background:#d94848;padding:18px 24px;text-align:center;">
              <h1 style="margin:0;color:#fff;font-size:24px;">REGISTRATION CONFIRMED</h1>
              <p style="margin:6px 0 0;color:#ffeaea;font-size:14px;">CELPIP Weekly Webinar</p>
            </td>
          </tr>
          <tr>
            <td style="padding:20px 22px;">
              <p style="margin:0 0 10px;font-size:18px;font-weight:700;color:#1d1d1d;">Dear ${customerName},</p>
              <p style="margin:0 0 14px;color:#2b2b2b;line-height:1.5;">Your payment has been processed and your webinar registration is confirmed.</p>
              <div style="background:#f3f3f3;border-left:4px solid #d94848;padding:12px;border-radius:5px;margin-bottom:14px;">
                <p style="margin:0 0 6px;font-size:12px;font-weight:700;letter-spacing:.8px;text-transform:uppercase;color:#333;">Join Your Webinar</p>
                <p style="margin:0;color:#d94848;font-size:16px;font-family:'Courier New',monospace;font-weight:700;word-break:break-all;">${GOOGLE_MEET_LINK}</p>
              </div>
              <p style="margin:0 0 8px;color:#333;">Saturday • 6:00 PM PST • 60 minutes • Google Meet</p>
              <p style="margin:0;color:#333;">Please join 5 minutes early for audio/video setup.</p>
            </td>
          </tr>
          <tr>
            <td style="background:#f4f1ef;padding:12px 22px;border-top:1px solid #e6ddd7;text-align:center;">
              <p style="margin:0 0 4px;font-size:14px;font-weight:700;color:#1f1f1f;">IELTS Corner</p>
              <p style="margin:0;font-size:12px;color:#666;">${gmailUser} | ieltscorner.ca</p>
            </td>
          </tr>
        </table>
      </td>
    </tr>
  </table>
</body>
</html>`;
}

function buildEmailText(customerName, gmailUser) {
  return `IELTS Corner — CELPIP Weekly Webinars

Registration Confirmed ✓

Hi ${customerName},

You're all set! Thank you for registering for this week's CELPIP webinar.

SESSION DETAILS
When: Saturday, 6:00 PM PST
Duration: 60 minutes
Format: Live on Google Meet

JOIN WEBINAR
${GOOGLE_MEET_LINK}

Please join 5 minutes early.

IELTS Corner
${gmailUser}
https://ieltscorner.ca`;
}

async function sendEmail({ gmailUser, gmailPassword, customerEmail, customerName }) {
  const transporter = nodemailer.createTransport({
    service: 'gmail',
    auth: {
      user: gmailUser,
      pass: gmailPassword,
    },
  });

  await transporter.sendMail({
    from: `IELTS Corner <${gmailUser}>`,
    to: customerEmail,
    subject: '✅ Your CELPIP Webinar Registration — Saturday 6:00 PM PST',
    text: buildEmailText(customerName, gmailUser),
    html: buildEmailHtml(customerName, gmailUser),
  });
}

export const handler = async (event) => {
  if (event.httpMethod !== 'POST') {
    return {
      statusCode: 405,
      body: JSON.stringify({ error: 'Method not allowed' }),
    };
  }

  const stripeApiKey = process.env.STRIPE_API_KEY;
  const webhookSecret = process.env.STRIPE_WEBHOOK_SECRET;
  const gmailUser = process.env.GMAIL_USER || 'kara.abdolmaleki@gmail.com';
  const gmailPassword = process.env.GMAIL_PASSWORD;

  if (!stripeApiKey || !webhookSecret || !gmailPassword) {
    return {
      statusCode: 500,
      body: JSON.stringify({ error: 'Server not configured: missing Stripe/Gmail env vars' }),
    };
  }

  const signature = event.headers['stripe-signature'] || event.headers['Stripe-Signature'];
  if (!signature) {
    return {
      statusCode: 400,
      body: JSON.stringify({ error: 'Missing Stripe signature' }),
    };
  }

  let stripeEvent;
  try {
    const stripe = new Stripe(stripeApiKey, { apiVersion: '2020-08-27' });
    stripeEvent = stripe.webhooks.constructEvent(event.body, signature, webhookSecret);

    if (stripeEvent.type !== 'checkout.session.completed') {
      return {
        statusCode: 200,
        body: JSON.stringify({ status: 'ignored', type: stripeEvent.type }),
      };
    }

    const session = stripeEvent.data.object;
    if (session.payment_status !== 'paid') {
      return {
        statusCode: 200,
        body: JSON.stringify({ status: 'ignored', reason: 'not_paid' }),
      };
    }

    const customerEmail = session?.customer_details?.email;
    const customerName = session?.customer_details?.name || 'there';

    if (!customerEmail) {
      return {
        statusCode: 400,
        body: JSON.stringify({ error: 'Missing customer email in checkout session' }),
      };
    }

    await sendEmail({
      gmailUser,
      gmailPassword,
      customerEmail,
      customerName,
    });

    return {
      statusCode: 200,
      body: JSON.stringify({ status: 'success', message: 'Confirmation email sent' }),
    };
  } catch (error) {
    return {
      statusCode: 400,
      body: JSON.stringify({ error: error?.message || 'Webhook processing failed' }),
    };
  }
};
